[Sema] [CSGen] Solve types through a constraint system

harlanhaskins opened the pull request:
This builds off work done by @CodaFi to implement a constraint solver that will solve for type variables.

CodaFi committed:
Introduce a type variable type

Keeping this here is fine, but having an interface to create type
variables that aren’t fresh is dangerous.

CodaFi committed:
Propagate types to closure expressions

CodaFi committed:
Parse shorthand for closure arguments

This currently enables

{ x, y, z, … in <stmts> }

and

{ <stmts> }

It does not yet handle

{ (x, y, z, …) -> T in <stmts> }

CodaFi committed:
Solve the type variables generated by closure expressions

CodaFi committed:
Propagate canonicalized types only

CodaFi committed:
Generate type variables in Sema only

CodaFi committed:
Typecheck subscripts

harlanhaskins committed:
Removed optionality for DataTypes. Implemented constraint generation for various expressions. Simplified solver code a bit.

harlanhaskins committed:
Ensure the trill_once declaration is not variadic.

harlanhaskins committed:
Make proper errors for unification failures.

harlanhaskins committed:
Always canonicalize types before solving.

harlanhaskins committed:
Added a `conforms` constraint

harlanhaskins committed:
Simplified and restructured constraint solver.

CodaFi commented on Sources/Semantic Analysis/ConstraintSolver.swift:
```diff
@@ -0,0 +1,139 @@
+      case (.bool, .int):
+        // Boolean values may coerce to integer values (but not vice-versa).
+        return [:]
+      case (_, .any), (.any, _):
```
Add an existential conversion expression and kill this with 🔥

harlanhaskins committed:
Uncommented varexpr code.

harlanhaskins committed:
Implement constraint gen for `is` and `as` expressions

harlanhaskins committed:
Don't store method names in the type checker's environment.

harlanhaskins committed:
Actually, don't store function names in the environment at all. The Context manages that.

harlanhaskins committed:
Removed bogus Int -> Bool solution and fix clang importer to add types to String macros.

segiddins commented on Sources/Base/Types.swift:
```diff
@@ -540,7 +540,11 @@ class TypeRefExpr: DeclRefExpr<TypeDecl> {
+    let expr = TypeRefExpr(type: self,
+                           name: Identifier(name: "\(self)", range: range),
+                           sourceRange: range)
+    expr.type = self
```
is this necessary, given we pass the type to the initializer?

harlanhaskins replied:
Good catch. I threw this in and forgot to remove it.

segiddins replied:
this never got removed...

harlanhaskins committed:
Removed metavariables and moved type variable pooling into the environment.

harlanhaskins committed:
Simplify solving constraints in the TypeChecker

harlanhaskins committed:
Move ConstraintSystem to its own type with its own methods. Explicitly substitute the types in constraints before solving.

harlanhaskins committed:
Clean up constraint solver errors.

harlanhaskins committed:
Fix TupleFieldLookupExpr constraint generation

segiddins commented on Sources/Base/ASTContext.swift:
```diff
@@ -281,24 +282,59 @@ public class ASTContext {
+    var missing = [MethodDecl]()
+    for method in methods {
+      var impl: MethodDecl?
+      for candidate in decl.methods(named: method.name.name) {
```
`impl = decl.methods(named:).first { ahveEqualSigs }`

harlanhaskins replied:
😄

segiddins commented on Sources/Base/ASTContext.swift:
```diff
@@ -281,24 +282,59 @@ public class ASTContext {
     case let expr as NumExpr:
       if case .int = canTy {
-        expr.type = contextualType
+        expr.type = canTy
```
didn't we want to continue propogating the sugared type?

harlanhaskins replied:
Oops.

CodaFi replied:
Ideally we propagate nothing and this entire method goes the way of the dinosaurs.

harlanhaskins replied:
Every other one of them propagates the `canTy`...

segiddins replied:
only as of this PR...

segiddins commented on Sources/Base/Functions.swift:
```diff
@@ -65,10 +65,20 @@ class FuncDecl: Decl { // func <id>(<id>: <type-id>) -> <type-id> { <expr>* }
     self.isPlaceholder = isPlaceholder
-    super.init(type: .function(args: args.map { $0.type }, returnType: returnType.type!),
+    var allValid = true
+    for arg in args {
```
`!args.contains {}`

harlanhaskins replied:
😄

CodaFi commented on Sources/Semantic Analysis/ConstraintGenerator.swift:
```diff
@@ -257,11 +257,12 @@ final class ConstraintGenerator: ASTTransformer {
-    system.constrainEqual(expr.decl!, lhsGoal)
-    let tau = env.freshTypeVariable()
+    guard case .tuple(let fields) = lhsGoal else {
+      return
```
`fatalError`?

harlanhaskins committed:
Don't bind the name of a VarAssignDecl while visiting its initializer.
Also fix constraints for static methods

segiddins committed:
[CS] Enumerate constraints in reverse

This way, eg for a ternary we unify to the ternary type first before subexpr types

harlanhaskins committed:
Don't emit any type variables for Array literals.

harlanhaskins committed:
Do emit type variables for arrays, but emit an explicit equal constraint to the array's field.

harlanhaskins committed:
Fix constraint generation for var assign decls where the explicit type is Any.

harlanhaskins committed:
Fix constraints for type metadata mirror generation.

harlanhaskins committed:
Fixed constraint resolution for static methods

harlanhaskins committed:
Added explicit master branch to LLVMSwift submodule

harlanhaskins committed:
Update remote submodule

harlanhaskins committed:
See what failed in the interpolation test

segiddins commented on Sources/Semantic Analysis/Sema.swift:
```diff
@@ -890,7 +879,7 @@ class Sema: ASTTransformer, Pass {
       }
     }
-
+    
```
whitespace

segiddins commented on trill.xcodeproj/project.xcworkspace/xcshareddata/trill.xcscmblueprint:
```diff
@@ -24,7 +24,7 @@
     },
     {
-      "DVTSourceControlWorkspaceBlueprintRemoteRepositoryURLKey" : "github.com:trill-lang\/trill.git",
+      "DVTSourceControlWorkspaceBlueprintRemoteRepositoryURLKey" : "github.com:CodaFi\/trill.git",
```
uh no

segiddins committed:
[Examples] Make failing output in string interpolation nicer

segiddins committed:
[LLVMWrappers] Set target options for debugging

segiddins committed:
[IRGen] Allow referencing self as a pointer in indirect initializers

harlanhaskins committed:
Flip goal and goalType for var assign decl constraints so the error reflects the actual issue

Otherwise, the error is:

let x: String = 100
                      ^ cannot convert String to Int

harlanhaskins committed:
Remove concatenation in string interpolation example

harlanhaskins committed:
Guard GC_INIT with a #if TRILL_ENABLE_GC

harlanhaskins committed:
Add concatenation back to example

harlanhaskins committed:
Fixed bug I introduced in the string interpolation test

harlanhaskins committed:
Removed concatenation again

harlanhaskins committed:
Removed errant ClangSwift entry

harlanhaskins committed:
Disable string interpolation test

harlanhaskins committed:
Add back string-interpolation test in broken directory
